### PR TITLE
authz: refactor providers init logic to their own packages

### PR DESCRIPTION
**Pull request by unknwon**
This PR is a prerequisite for merging https://github.com/sourcegraph/sourcegraph/pull/6427 because it fixes the cyclic import compile error.

Test plan: CI

### CODE REVIEW

**Comment by tsenart** [comment]
Two thoughts before reviewing this further:

- What specific cyclic import is being broken up with this change?
- In general, I'd like to see less package level `init` functions, less global state mutation, less registration patterns and more dependency injection.

**Comment by unknwon** [comment]
> * What specific cyclic import is being broken up with this change?

In the context of #6427, without this refactoring, we would have  `authz/authz_bitbucketserver.go` imports `authz/bitbucketserver/*` which imports `authz/permissions.go`.

Another way considered but I decided not go with is put permission types in a subpackage `authz/permissions`. I thought `permissions.UserPermissions` is redundant in type name, and `permissions.User` feels somewhat stupid from a non-native speaker's POV. 😂 A type with the name `authz.UserPermissions` makes more sense to me.

> * In general, I'd like to see less package level `init` functions, less global state mutation, less registration patterns and more dependency injection.

- Are there previous discussions I could take a look about when we should use which pattern, or in favor of some patterns?
- In this particular case, would you mind explain why the registration pattern is inferior to the dependency injection?
- Any links for best examples you think that use dependency injection in our codebase would be very much appreciated! (because I often confused by what you mean exactly, when we mentioned DI during previous syncs, meetings.)

**Comment by tsenart** [comment]
I’m writing on iPhone and about to board a flight so apologies for typos or terseness.

Regarding the breaking of the import cycle, got it, thanks for the explanation.

For examples of dependency injection, see most of the code in repo-updater. Read this as a primer on Go programming without global state:

- https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html
- https://twitter.com/peterbourgon/status/1133492111012184064?s=20

**Comment by unknwon** [comment]
I think I just fall in love with "A theory of modern Go", I'm going to try this pattern in my side project over the weekend and make changes to this PR.

**Comment by unknwon** [comment]
All changes in this PR is moving code around with:
1. some variable rename
2. and remove unused variables

In an overview, code in `internal/authz` either:
1. move up to `enterprise/cmd/frontend` (where they're used)
2. or move down to `internal/authz/bitbucketserver`, `internal/authz/github`, `internal/authz/gitlab`.

Please take a look now @tsenart!

File: enterprise/cmd/frontend/internal/authz/bitbucketserver/authz.go [hunk 1]
```diff
@@ -1,66 +1,62 @@
...
+// to false. "Warnings" are all other validation problems.
+func NewAuthzProviders(
 	conns []*schema.BitbucketServerConnection,
-) (
-	authzProviders []authz.Provider,
-	seriousProblems []string,
-	warnings []string,
-) {
+	db *sql.DB,
+) (ps []authz.Provider, problems []string, warnings []string) {
```
Comment (tsenart) [inline enterprise/cmd/frontend/internal/authz/bitbucketserver/authz.go]:
```suggestion
) (ps []authz.Provider, problems, warnings []string) {
```

**Pull request merged**